ldap/servers/slapd/auditlog.c

```diff
@@ -156,22 +160,116 @@ write_auditfail_log_entry(Slapi_PBlock *pb)
+        write_audit_file(NULL, SLAPD_AUDIT_LOG, operation_get_type(op), dn, change, flag, curtime, pbrc, SLAPD_AUDITFAIL_LOG);
     } else {
         /* If we have our own auditfail log path */
-        write_audit_file(SLAPD_AUDITFAIL_LOG, operation_get_type(op), dn, change, flag, curtime, pbrc, SLAPD_AUDITFAIL_LOG);
+        write_audit_file(NULL, SLAPD_AUDITFAIL_LOG, operation_get_type(op), dn, change, flag, curtime, pbrc, SLAPD_AUDITFAIL_LOG);
     }
     slapi_ch_free_string(&auditfail_config);
     slapi_ch_free_string(&audit_config);
 }
 
+/*
+ * Write "requested" attributes from the entry to the audit log as "comments"
+ *
+ *   Slapi_Entry *entry - the entry being updated
+ *   lenstr *l - the audit log buffer
+ *
+ *   Resulting output in the log:
+ *
+ *       #ATTR: VALUE
+ *       #ATTR: VALUE
+ */
+static void
+add_entry_attrs(Slapi_Entry *entry, lenstr *l)
+{
+    Slapi_Attr *entry_attr = NULL;
+    char *display_attrs = NULL;
+    char *req_attr = NULL;
+    char *last = NULL;
+
+    if (entry == NULL) {
+        /* auditfail log does not have an entry to read */
+        return;
+    }
+
+    display_attrs = config_get_auditlog_display_attrs();
+    if (display_attrs == NULL) {
+        return;
+    }
+
+    entry_attr = entry->e_attrs;
+    if (strcmp(display_attrs, "*")) {
+        /* Return specific attributes */
+        for (req_attr = ldap_utf8strtok_r(display_attrs, ", ", &last); req_attr;
+             req_attr = ldap_utf8strtok_r(NULL, ", ", &last))
+        {
+            char **vals = slapi_entry_attr_get_charray(entry, req_attr);
+            for(size_t i = 0; vals && vals[i]; i++) {
+                char log_val[256] = {0};
+
+                if (strlen(vals[i]) > 256) {
+                    strncpy(log_val, vals[i], 252);
+                    strcat(log_val, "...");
+                } else {
+                    strcpy(log_val, vals[i]);
+                }
+                addlenstr(l, "#");
+                addlenstr(l, req_attr);
+                addlenstr(l, ": ");
+                addlenstr(l, log_val);
+                addlenstr(l, "\n");
+            }
+        }
+    } else {
+        /* Return all attributes */
+        for (; entry_attr; entry_attr = entry_attr->a_next) {
+            Slapi_Value **vals = attr_get_present_values(entry_attr);
+            char *attr = NULL;
+            const char *val = NULL;
+
+            slapi_attr_get_type(entry_attr, &attr);
+            if (strcmp(attr, PSEUDO_ATTR_UNHASHEDUSERPASSWORD) == 0) {
+                /* Do not write the unhashed clear-text password */
+                continue;
+            }
+
+            if (strcasecmp(attr, SLAPI_USERPWD_ATTR) == 0) {
+                /* userpassword - mask the value */
+                addlenstr(l, "#");
+                addlenstr(l, attr);
+                addlenstr(l, ": *********************************\n");
```

Comment:
If we do `userPassword` this way, should we also process `nsslapd-rootpw` similarly? 

And just to confirm, is it by design that we don't replace the userPassword with `****` when userPassword is requested directly? (i.e. `nsslapd-auditlog-display-attrs: userPassword cn`)

Reply:
Yes if you are explicitly asking for an attribute then you get its value.  Doesn't make sense to ask for it and then the server hide the value.  

And yes I will add nsslapd-rootpw